boot/freeldr/freeldr/arch/realmode/amd64.S

```diff
@@ -24,13 +24,29 @@ Startup:
     mov gs, ax
     mov ss, ax
 
+    /* Don't clear the necessary variables if it's not the first boot */
+    cmp byte ptr ds:[FirstBoot], 0
+    jz .IsNotTheFirstBoot
+
+    /* Clear the necessary variables */
+    xor eax, eax
+    mov byte ptr ds:[BSS_BootedViaNtldr], al
+    mov dword ptr ds:[BSS_NtldrAddress], eax
+    mov dword ptr ds:[BSS_NtldrAddress + 4], eax
+    mov byte ptr ds:[FirstBoot], al
```

Comment:
It seems that all this thing could be moved into the `checkNtldrBoot` function:
1. Because all the code presently between `.IsNotTheFirstBoot:` and the `call checkNtldrBoot` is executed in any case (true here and in the i386.S file);
2. This check-for-frst-boot and clearing code appears to be done also in the i386.S file;
3. It would allow to make the usage of the relocateFreeldr-related variables a bit more localized.

If my observations are correct, then the `checkNtldrBoot` function would become:
```asm
checkNtldrBoot:
    /* Don't clear the necessary variables if it's not the first boot */
    cmp byte ptr ds:[FirstBoot], 0
    jz .IsNotTheFirstBoot

    /* Clear the necessary variables */
    xor eax, eax
    mov byte ptr ds:[BSS_BootedViaNtldr], al
    mov dword ptr ds:[BSS_NtldrAddress], eax
    mov dword ptr ds:[BSS_NtldrAddress + 4], eax
    mov byte ptr ds:[FirstBoot], al

.IsNotTheFirstBoot
    /* Check if it's booted as NTLDR */
    mov ax, cs
    cmp ax, NTLDR_BASE / 16
    jae relocateFreeldr

    ret
```



```suggestion
```

Reply:
Yes the change is correct!